core/src/main/java/io/undertow/server/handlers/PathTemplateHandler.java

```diff
@@ -20,85 +20,155 @@
+     *
+     * @param rewriteQueryParameters Path parameters that are returned by the underlying router will be added as
+     *                               query parameters to the exchange if this flag is 'true'.
+     */
+    public PathTemplateHandler(final boolean rewriteQueryParameters) {
         this(ResponseCodeHandler.HANDLE_404, rewriteQueryParameters);
     }
 
-    public PathTemplateHandler(HttpHandler next) {
+    /**
+     * Sets 'rewriteQueryParameters' to 'true'.
+     *
+     * @param next The next (default) handler to use when requests do not match any of the specified
+     *             templates.
+     */
+    public PathTemplateHandler(final HttpHandler next) {
         this(next, true);
     }
 
-    public PathTemplateHandler(HttpHandler next, boolean rewriteQueryParameters) {
+    /**
+     * @param next                   The next (default) handler to use when requests do not match any of the specified
+     *                               templates.
+     * @param rewriteQueryParameters Path parameters that are returned by the underlying router will be added as
+     *                               query parameters to the exchange if this flag is 'true'.
+     */
+    public PathTemplateHandler(final HttpHandler next, final boolean rewriteQueryParameters) {
+        Objects.requireNonNull(next);
+
         this.rewriteQueryParameters = rewriteQueryParameters;
-        this.next = next;
+        builder = PathTemplateRouter.SimpleBuilder.newBuilder(next);
+        router = builder.build();
     }
 
-    @Override
-    public void handleRequest(HttpServerExchange exchange) throws Exception {
-        PathTemplateMatcher.PathMatchResult<HttpHandler> match = pathTemplateMatcher.match(exchange.getRelativePath());
-        if (match == null) {
-            next.handleRequest(exchange);
-            return;
-        }
-        exchange.putAttachment(PATH_TEMPLATE_MATCH, new PathTemplateMatch(match.getMatchedTemplate(), match.getParameters()));
-        exchange.putAttachment(io.undertow.util.PathTemplateMatch.ATTACHMENT_KEY, new io.undertow.util.PathTemplateMatch(match.getMatchedTemplate(), match.getParameters()));
-        if (rewriteQueryParameters) {
-            for (Map.Entry<String, String> entry : match.getParameters().entrySet()) {
-                exchange.addQueryParam(entry.getKey(), entry.getValue());
-            }
+    /**
+     * Adds a template and handler to the underlying router.
+     *
+     * @param uriTemplate The URI path template.
+     * @param handler     The handler to use for requests that match the specified template.
+     *
+     * @return Reference to this handler.
+     */
+    public PathTemplateHandler add(final String uriTemplate, final HttpHandler handler) {
+        Objects.requireNonNull(uriTemplate);
+        Objects.requireNonNull(handler);
+
+        // Router builders are not thread-safe, so we need to synchronize.
```

Comment:
Just a quick glance:
1.Why would the need to be?
2. in which cases this would be needed?
3. Im not sure if such design of builder is optimal. AFAIR in other builders are context one-offs
4. AFAIR, in general, if there is no need to sync, we dont do it, rather rely on local var/single failure.

Reply:
The short answer is for backwards compatibility. The original PathTemplateHandler class (also PathTemplateRouter) provided synchronised mutation of the underlying routes.

The longer answer:
1. The underlying router itself is immutable and therefore thread-safe. The builder that creates the router is not thread-safe as it is intended to be used from a single thread to instantiate a router. For most use cases that I can think of the routes (path templates) should be discovered/added during start-up of a service.  Once discovered, a router can be built and then used to serve many many requests.  Therefore, I believe that in most services instantiating a router is a once-off process / invocation vs probably millions of invocations of the route method.  For this reason I believe that overall performance is served well by a slightly more expensive (computationally) instantiation process vs a very cheap routing process. Yet, this does not mean that calling the builder must be synchronised.... in my opinion most use cases must instantiate a router and pass it to an immutable handler such as the new PathTemplateRouterHandler class - which I provided as an example. We can add/offer builders for handlers - for convenience sake - that do the same thing as PathTemplateHandler and PathTemplateRouter, but that create immutable instances of those handlers instead of the current implementations that support mutating the underlying routers. This would however be a breaking change for developers who use and mutate the current implementations, so my recommendation would be to leave the current implementations as they are - for backwards compatibility - to deprecate them and to offer the new pattern as new, immutable handler classes that developers can switch to.
2. I believe there are very very few use cases that require synchronised mutation of the routing handlers. For those use cases, developers should implement their own synchronised mutation of underlying handlers. As I have mentioned before, I just implemented the synchronised mutation to be consistent with the previous implementations in order to avoid introducing breaking changes.
3. I agree that this is not optimal, but I am not sure if we are prepared to introduce this as a breaking change?
4. Agreed. In this case there should be no need to sync for 99%+ of use cases.

As a last note - in case it is not that obvious from the code and in case we want to keep the backwards compatibility - the synchronisation only happens when the underlying routers are modified, so when new path templates are added.  It only synchronises the mutations amongst multiple threads trying to mutate the routes concurrently.  The actual routing of requests are never synchronised - not even when other threads are busy mutating the routers.  For this reason I believe that it is okay to leave the synchronisation in order to maintain backwards compatibility. And then perhaps to deprecate and offer alternative classes in a next PR?